src/pages/studyView/charts/ChartContainer.tsx

```diff
@@ -90,50 +91,122 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
             return document.createElementNS("http://www.w3.org/2000/svg", "svg");
         }
     }
+    @computed get chartFilters(){
+        let filters = _.filter(this.props.filter.clinicalDataEqualityFilters, filter => _.isEqual(filter.clinicalDataType+'_'+filter.attributeId, this.props.chartMeta.uniqueKey))
+        return _.isEmpty(filters) ? [] : filters[0].values;
+    }
 
     @computed get showPieControlIcon(){
-        return this.props.clinicalAttribute.datatype==='STRING' && this.props.chartType === ChartType.TABLE;
+        return this.props.chartMeta.clinicalAttribute.datatype==='STRING' && this.props.chartMeta.chartType === ChartType.TABLE;
     }
 
     @computed get showTableControlIcon(){
-        return this.props.chartType === ChartType.PIE_CHART && this.props.clinicalAttribute.datatype==='STRING';
+        return this.props.chartMeta.chartType === ChartType.PIE_CHART && this.props.chartMeta.clinicalAttribute.datatype==='STRING';
+    }
+
+    private readonly data = remoteData<ClinicalDataCountWithColor[]>({
+        await: () => [this.initialData],
+        invoke: async () => {
+             return this.props.dataCache.get({ attribute: this.props.chartMeta.clinicalAttribute, filters: this.props.filter }).result!
+        },
+        default: []
+    })
+
+    private readonly initialData = remoteData<ClinicalDataCountWithColor[]>({
+        invoke: async () => {
+            return this.props.dataCache.get({ attribute: this.props.chartMeta.clinicalAttribute, filters: {studyIds: this.props.filter.studyIds} as any }).result!;
+        },
+        default: [],
+        onResult:(result) => {
+            if (result.length > 20) {
+                this.props.changeChartType(this.props.chartMeta.uniqueKey, ChartType.TABLE, true);
+            }
+        }
+    })
+
+
+
+    @computed get chartWidth(){
+        let chartWidth = styles.chartWidthTwo
+        if(this.props.chartMeta.chartType === ChartType.PIE_CHART){
```

Comment:
how many different chart types are there?

Reply:
there are 5.

Reply:
seems like then are going to have a lot of switch statements and they need to have all five options.  hard to maintain.  would it be better to have subclasses which implement these members?